packages/flutter_tools/lib/src/commands/multi_drive.dart

```diff
@@ -0,0 +1,375 @@
+  // Try to install and start apps in parallel
+  List<Future<LaunchResult>> installAndStartAppFunctions = <Future<LaunchResult>>[];
+  // Iterate through device ids to get devices
+  for(String deviceID in devices.keys) {
+    Map<String, String> config = devices[deviceID];
+
+    String mainPath = findMainDartFile(config['app-path']);
+
+    if (await fs.type(mainPath) != FileSystemEntityType.FILE) {
+      printError('Tried to run $mainPath, but that file does not exist.');
+      return 1;
+    }
+
+    Device device = findDevice(command.devices, deviceID);
+
+    // TODO(devoncarew): We should remove the need to special case here.
+    if (device is AndroidDevice) {
+      printTrace('Building an APK.');
+      int result = await build_apk.buildApk(
+        device.platform,
+        target: config['app-path'], //command.target,
+        buildMode: command.getBuildMode()
+      );
+
+      if (result != 0)
+        return result;
+    }
+
+    printTrace('Stopping previously running application, if any.');
+    await appsStopper(command.applicationPackages, device);
+
+    installAndStartAppFunctions.add(makeInstallAndStartApp(command, device, mainPath, config['debug-port']));
+  }
+  // command.specs['devices'].forEach((String deviceID, Map<String, String> value) async {
+  //
+  // });
+  // Install and start apps in parallel
+  // TODO: modify internal synchronous calls to asynchronous calls
+  Future.wait(installAndStartAppFunctions)
+        .then((List<LaunchResult> results) {
+          for(LaunchResult result in results) {
+            if(!result.started) return 2;
+          }
+        });
+  return 0;
+}
+
+Future<LaunchResult> makeInstallAndStartApp(
+  MultiDriveCommand command,
+  Device device,
+  String mainPath,
+  String debugPort) async {
+    printTrace('Installing application package.');
+    ApplicationPackage package = command.applicationPackages
+        .getPackageForPlatform(device.platform);
+    if (device.isAppInstalled(package))
+      device.uninstallApp(package);
+    device.installApp(package);
+
+    Map<String, dynamic> platformArgs = <String, dynamic>{};
+    if (command.traceStartup)
+      platformArgs['trace-startup'] = command.traceStartup;
+
+    printTrace('Starting application.');
+    LaunchResult result = await device.startApp(
+      package,
+      command.getBuildMode(),
+      mainPath: mainPath,
+      route: command.route,
+      debuggingOptions: new DebuggingOptions.enabled(
+        command.getBuildMode(),
+        startPaused: true,
+        observatoryPort: int.parse(debugPort)
+      ),
+      platformArgs: platformArgs
+    );
+    return result;
+}
+
+/// Runs driver tests.
+typedef Future<int> MultiDeviceTestRunner(List<String> testArgs);
+MultiDeviceTestRunner testRunner = runMultiDeviceTests;
+void restoreMultiDeviceTestRunner() {
+  testRunner = runMultiDeviceTests;
+}
+
+Future<int> runMultiDeviceTests(List<String> testArgs) async {
```

Comment:
Same here. just `runTests` would be better, because we're actually running a single test script. Consider directly calling the `runTests` defined in `driver.dart`, if possible.